### PR TITLE
Types: Adds type safety to appEvents

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Add type safety to appEvents, ensuring that the event exists, and that they associated payload conforms to a specified interface. (Similar to redux actions)

### CODE REVIEW

File: packages/grafana-data/src/types/events.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  role?: string;
+  shortcut?: string;
+}
+
+export interface AppEvent<T> {
+  readonly name: string;
+  payload?: T;
+}
+
+export function eventFactory<T>(name: string): AppEvent<T> {
```
Comment (aocenas) [inline packages/grafana-data/src/types/events.ts]: One thing that would probably make this a bit more safe is to check whether the names are unique. I think we already do that in redux stores with actions which can also clash.

Comment (kaydelaney) [reply packages/grafana-data/src/types/events.ts]: That's a good point! I'll implement that

**Comment by ryantxu** [comment]
This is super exciting to see.  The untyped events all over angular panels are mysterious!

I would love to see how we could broadcast/process them with rxjs rather than Emitter -- that will make more sense for events like the shared tooltip and annotation events

**Review by hugohaggmark — COMMENTED** [review]
Awesome work!

File: packages/grafana-data/src/types/events.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+  role?: string;
+  shortcut?: string;
+}
+
+export interface AppEvent<T> {
+  readonly name: string;
+  payload?: T;
+}
+
+const typeList: Set<string> = new Set();
```
[the remark this answers is not in the document]
Comment (kaydelaney) [reply packages/grafana-data/src/types/events.ts]: https://github.com/grafana/grafana/pull/19433

File: public/app/plugins/datasource/elasticsearch/specs/datasource.test.ts [hunk 3]
```diff
@@ -8,16 +8,17 @@ import { TimeSrv } from 'app/features/dashboard/services/TimeSrv';
...
   const backendSrv: any = {
     datasourceRequest: jest.fn(),
   };
 
-  const $rootScope: any = {
+  const $rootScope: GrafanaRootScope = {
     $on: jest.fn(),
     appEvent: jest.fn(),
-  };
+  } as any;
```
Comment (hugohaggmark) [inline public/app/plugins/datasource/elasticsearch/specs/datasource.test.ts]: Nitpick: Is there a way to get rid of the `any` here?

File: public/app/plugins/datasource/stackdriver/datasource.ts [hunk 4]
```diff
@@ -287,7 +288,8 @@ export default class StackdriverDatasource extends DataSourceApi<StackdriverQuer
 
       return this.metricTypes;
     } catch (error) {
-      appEvents.emit('ds-request-error', this.formatStackdriverError(error));
+      //@ts-ignore
```
Comment (hugohaggmark) [inline public/app/plugins/datasource/stackdriver/datasource.ts]: Nitpick: why the @ts-ignore and can we remove it somehow?

**Review by hugohaggmark — APPROVED** [review]
Fantastic work with this, I found some minor things that you could choose to fix before merging. I've tested some but I certainly haven't covered it all. 

Again, great work!

**Review by aocenas — APPROVED** [review]
👍 Clicked around a bit and did not find any issue.

File: public/app/core/services/keybindingSrv.ts [hunk 5]
```diff
@@ -164,37 +181,37 @@ export class KeybindingSrv {
     this.$location.search(search);
   }
 
-  setupDashboardBindings(scope: any, dashboard: any) {
+  setupDashboardBindings(scope: IRootScopeService & AppEventEmitter, dashboard: any) {
```
Comment (hugohaggmark) [inline public/app/core/services/keybindingSrv.ts]: Nice!

File: public/app/plugins/datasource/stackdriver/datasource.ts [hunk 6]
```diff
@@ -287,7 +288,7 @@ export default class StackdriverDatasource extends DataSourceApi<StackdriverQuer
 
       return this.metricTypes;
     } catch (error) {
-      appEvents.emit('ds-request-error', this.formatStackdriverError(error));
+      appEvents.emit(dsRequestError, { error: { data: { error: this.formatStackdriverError(error) } } });
```
Comment (hugohaggmark) [inline public/app/plugins/datasource/stackdriver/datasource.ts]: Just noticed this one, out of curiosity why this change? Why not add this `error` to the `this. formatStackdriverError `?

File: public/app/routes/GrafanaCtrl.ts [hunk 7]
```diff
@@ -27,14 +27,43 @@ import { UtilSrv } from 'app/core/services/util_srv';
...
+import { ILocationService, ITimeoutService, IRootScopeService, IAngularEvent } from 'angular';
+
+export interface AppEventEmitter {
+  /**
+   * DEPRECATED.
+   */
+  appEvent(name: string, data?: any): void;
+
+  appEvent<T extends undefined>(event: AppEvent<T>): void;
+  appEvent<T extends Partial<T> extends T ? Partial<T> : never>(event: AppEvent<T>): void;
```
Comment (hugohaggmark) [inline public/app/routes/GrafanaCtrl.ts]: Maybe just add a small comment explaining the Partial type api here?

File: public/app/routes/dashboard_loaders.ts [hunk 8]
```diff
@@ -4,18 +4,20 @@ import { UrlQueryMap } from '@grafana/runtime';
...
 import { BackendSrv } from 'app/core/services/backend_srv';
 import { ILocationService } from 'angular';
+import { AppEventEmitter } from './GrafanaCtrl';
+import { dashboardFetchStart } from '@grafana/data';
 
 export class LoadDashboardCtrl {
   /** @ngInject */
   constructor(
-    $scope: any,
+    $scope: AppEventEmitter & { [key: string]: any },
```
Comment (hugohaggmark) [inline public/app/routes/dashboard_loaders.ts]: I think I've seen `{ [key: string]: any }` type a couple of times in this PR, maybe we should go ahead and create a `Scope` type?

File: public/app/plugins/datasource/stackdriver/datasource.ts (hunk 6 above)
Comment (kaydelaney) [reply public/app/plugins/datasource/stackdriver/datasource.ts]: Made this change since the handler for `dsRequestError` events (`QueryInspector > onDataSourceResponse`) doesn't deal with strings (which was previously being sent), so I changed it so that it's sent an object it can properly deal with. I didn't change what `formatStackdriverError` returns since elsewhere where it's used it _does_ expect a string. Not a replacement for proper typing, but it's a quick fix until we have well-defined error types.

Comment (hugohaggmark) [reply public/app/plugins/datasource/stackdriver/datasource.ts]: Ok, thanks for clarification

File: packages/grafana-data/src/types/events.ts [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+export const elasticQueryUpdated = eventFactory('elastic-query-updated');
+
+export const layoutModeChanged = eventFactory<string>('layout-mode-changed');
+
+export const jsonDiffReady = eventFactory('json-diff-ready');
+
+export const closeTimepicker = eventFactory('closeTimepicker');
+
+export const routeUpdated = eventFactory('$routeUpdate');
```
Comment (torkelo) [inline packages/grafana-data/src/types/events.ts]: Almost all of these are internal app events that we do not want to expose to plugins (and are legacy mechanics that we want to move away from over time). So most of these should be moved to app/types/events.ts (start with all until we know which are needed to be used from plugins).

File: packages/grafana-ui/src/types/events.ts [hunk 10]
```diff
@@ -0,0 +1,6 @@
+import { eventFactory } from '@grafana/data';
+import { DataQueryResponseData, DataQueryError } from '.';
+
+export const dataReceived = eventFactory<DataQueryResponseData[]>('data-received');
+export const dataError = eventFactory<DataQueryError>('data-error');
+export const dataSnapshotLoad = eventFactory<DataQueryResponseData[]>('data-snapshot-load');
```
Comment (torkelo) [inline packages/grafana-ui/src/types/events.ts]: there is only single namespace for @grafana/ui, and these are only for old angular plugins and something we want to migrate away from. 

maybe we can add a legacy prefix?

Comment (torkelo) [reply packages/grafana-ui/src/types/events.ts]: or `panelEvent` like panelEventDataReceived

**Review by torkelo — CHANGES_REQUESTED** [review]
Great job!  I think we need to move the events away from @grafana/data and in to app/types, and then only the ones that are needed can be in @grafana/runtime  but with a prefix lik "appEvent", as we only have a single namespace for these packages we need to be a bit restrive on short exported names.

File: packages/grafana-data/src/types/events.ts (hunk 9 above)
Comment (kaydelaney) [reply packages/grafana-data/src/types/events.ts]: Sure, will do

File: packages/grafana-ui/src/types/events.ts (hunk 10 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/types/events.ts]: Could hide them behind an `AppEvents` namespace?

Comment (torkelo) [reply packages/grafana-ui/src/types/events.ts]: what do you mean? (we can't use typescript namespaces, they will be deprecated I think). 

Do you mean exporting an object named AppEvents with static properties? 

like 
```
export const AppEvents = {
  dataReceived: ...
}```

Think it would be PanelEvents =  ..
 in that case.

Comment (kaydelaney) [reply packages/grafana-ui/src/types/events.ts]: Also, since we want to move events away from @grafana/data and to move those that are needed to @grafana/runtime, would that mean introducing a dependency on @grafana/runtime in @grafana/ui, since the ui package has its own events and depends on `eventFactory`? Or do you think it would be better to have `eventFactory` in @grafana/data since @grafana/ui already has a dependency on that package

Comment (torkelo) [reply packages/grafana-ui/src/types/events.ts]: ah, did not think about that. @grafana/ui should not have a dependency on @grafana/runtime. 

The DataQueryXXXX types should really be in @grafana/data we have just not moved them yet. But for now let's keep the eventFactory in @grafana/data.

File: public/app/features/dashboard/state/PanelModel.ts [hunk 11]
```diff
@@ -5,11 +5,25 @@ import { Emitter } from 'app/core/utils/emitter';
...
+  panelInitialized,
+  panelTeardown,
+  clientRefreshed,
+  eventFactory,
+} from '@grafana/data';
 
 import config from 'app/core/config';
 
 import { PanelQueryRunner } from './PanelQueryRunner';
+import { rendered } from 'app/types';
```
Comment (torkelo) [inline public/app/features/dashboard/state/PanelModel.ts]: rendered? need some prefix or suffix here to make this name more clear.

File: public/app/types/events.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import TimeSeries from 'app/core/time_series2';
+import TableModel from 'app/core/table_model';
+import { eventFactory } from '@grafana/data';
+import { DashboardModel } from 'app/features/dashboard/state';
+
+export const rendered = eventFactory<Partial<TableModel | TimeSeries[]>>('render');
```
Comment (torkelo) [inline public/app/types/events.ts]: rename to panelRenderEvent ? or panelEventRender  (or namespace it inside an object along all the PanelEvents). Should not be rendered as this is not something that happens after render but an event to initiate rendering

File: packages/grafana-data/src/types/events.ts [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+export function eventFactory<T = undefined>(name: string): AppEvent<T> {
+  if (typeList.has(name)) {
+    throw new Error(`There is already an event defined with type '${name}'`);
+  }
+
+  typeList.add(name);
+  return { name };
+}
+
+export const alertWarning = eventFactory<AlertPayload>('alert-warning');
```
Comment (torkelo) [inline packages/grafana-data/src/types/events.ts]: I am not a big fan of executing functions in global scope (when bundles load) but sometimes it ok. But I wonder if we can solve the namespacing problem and that in one go. 

Probably a bad idea (not sure): 

```
export class AppEvents {
   
   alertWarning() { return eventFactory<AlertPayload>('alert-warning') }
}

File: public/app/routes/GrafanaCtrl.ts [hunk 14]
```diff
@@ -27,14 +27,44 @@ import { UtilSrv } from 'app/core/services/util_srv';
...
+  playlistStopped,
+  toggleKioskMode,
+  toggleViewMode,
+  hideDashSearch,
+  alertSuccess,
+  AppEvent,
+} from '@grafana/data';
+import { ILocationService, ITimeoutService, IRootScopeService, IAngularEvent } from 'angular';
+
+export interface AppEventEmitter {
```
Comment (torkelo) [inline public/app/routes/GrafanaCtrl.ts]: Maybe we can move this to app/types

**Review by torkelo — COMMENTED** [review]
Some additional comments :)

File: packages/grafana-ui/src/types/events.ts (hunk 10 above)
Comment (ryantxu) [reply packages/grafana-ui/src/types/events.ts]: `@grafana/data` makes sense to me -- we could use somethign like the approach we have for dateMath to give things a "namespace"

https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/datetime/index.ts#L6

Comment (kaydelaney) [reply packages/grafana-ui/src/types/events.ts]: Yes, I meant a namespace in the sense of an object with the the events as properties, sorry for the confusion. Would that work for you, @torkelo ?

File: public/app/types/events.ts (hunk 12 above)
Comment (kaydelaney) [reply public/app/types/events.ts]: Thanks for pointing this out. Since it's a panel event does that suggest it should belong in @grafana/ui as well? If so I suppose TableModel/TimeSeries would need to be moved to one of the external packages as well

File: packages/grafana-data/src/types/events.ts (hunk 13 above)
Comment (kaydelaney) [reply packages/grafana-data/src/types/events.ts]: Not sure how this would work with how duplicate event names are disallowed, since invoking, say, `alertWarning` more than once would throw an error since you're basically registering the event each time you call `alertWarning`. I thought maybe memoizing calls to eventFactory would fix this but then that would stop eventFactory from throwing if someone really did accidentally try to register two events with the same name.
One (probably very bad) idea would be to do something like this:
```ts
export class AppEvents {
   
   alertWarning() {
      const alertWarningEvent = eventFactory<AlertPayload>('alert-warning');
      AppEvents.prototype.alertWarning = () => alertWarningEvent;
      return alertWarningEvent;
   }
} 
```
which would return the same value on subsequent calls, but also throw if 'alert-warning' were to be re-registered elsewhere. Probably not very good practice though

Comment (kaydelaney) [reply packages/grafana-data/src/types/events.ts]:
```ts
export class AppEvents {
   
   get alertWarning() {
      const alertWarningEvent = eventFactory<AlertPayload>('alert-warning');
      Object.defineProperty(this, "alertWarning", {
        get: () => alertWarning
      };
      return alertWarningEvent;
   }
} 
```

Comment (torkelo) [reply packages/grafana-data/src/types/events.ts]: Maybe it's going to be easier / cleaner to declare these in global scope. If we limit the number defined in @grafana/data to only the essential ones used by plugins (alerts?), the reset in app/core. Think these will be fewer over time as we migrate more code to react

File: packages/grafana-ui/src/types/events.ts (hunk 10 above)
Comment (torkelo) [reply packages/grafana-ui/src/types/events.ts]: Think these PanelEvents needs to be defined in @grafana/ui until we move the DataQueryResponse types

Comment (torkelo) [reply packages/grafana-ui/src/types/events.ts]: > meant a namespace in the sense of an object with the the events as properties

that will work :+1:

File: packages/grafana-data/src/types/events.ts (hunk 13 above)
Comment (kaydelaney) [reply packages/grafana-data/src/types/events.ts]: Sounds good

File: packages/grafana-data/src/types/index.ts [hunk 15]
```diff
@@ -13,3 +13,4 @@ export * from './graph';
 export * from './ScopedVars';
 export * from './transformations';
 export * from './vector';
+export * from './events';
```
Comment (torkelo) [inline packages/grafana-data/src/types/index.ts]: was hoping we would namespace these as well { AppEvents }

File: public/app/features/panel/metrics_panel_ctrl.ts [hunk 16]
```diff
@@ -42,8 +43,8 @@ class MetricsPanelCtrl extends PanelCtrl {
     this.scope = $scope;
     this.panel.datasource = this.panel.datasource || null;
 
-    this.events.on('refresh', this.onMetricsPanelRefresh.bind(this));
-    this.events.on('panel-teardown', this.onPanelTearDown.bind(this));
+    this.events.on(clientRefreshed, this.onMetricsPanelRefresh.bind(this));
+    this.events.on(panelTeardown, this.onPanelTearDown.bind(this));
```
Comment (torkelo) [inline public/app/features/panel/metrics_panel_ctrl.ts]: these are also PanelEvents.

File: public/app/features/panel/panel_ctrl.ts [hunk 17]
```diff
@@ -56,11 +59,11 @@ export class PanelCtrl {
       this.pluginName = plugin.name;
     }
 
-    $scope.$on('component-did-mount', () => this.panelDidMount());
+    $scope.$on(componentDidMount.name, () => this.panelDidMount());
   }
 
   panelDidMount() {
-    this.events.emit('component-did-mount');
+    this.events.emit(componentDidMount);
```
Comment (torkelo) [inline public/app/features/panel/panel_ctrl.ts]: this should also be on PanelEvents.

File: public/app/features/panel/panel_ctrl.ts [hunk 18]
```diff
@@ -193,7 +196,7 @@ export class PanelCtrl {
       click: 'ctrl.editPanelJson(); dismiss();',
     });
 
-    this.events.emit('init-panel-actions', menu);
+    this.events.emit(initPanelActions, menu);
```
Comment (torkelo) [inline public/app/features/panel/panel_ctrl.ts]: this is also a PanelEvents event

File: public/app/features/panel/panel_ctrl.ts [hunk 19]
```diff
@@ -211,8 +214,8 @@ export class PanelCtrl {
     this.height = calculateInnerPanelHeight(this.panel, containerHeight);
   }
 
-  render(payload?: any) {
-    this.events.emit('render', payload);
+  render(payload?: TableModel | TimeSeries[]) {
+    this.events.emit(panelEventRender, payload);
```
Comment (torkelo) [inline public/app/features/panel/panel_ctrl.ts]: this is also a PanelEvents event

File: public/app/features/panel/panel_directive.ts [hunk 20]
```diff
@@ -102,15 +104,15 @@ module.directive('grafanaPanel', ($rootScope, $document, $timeout) => {
...
+      ctrl.events.on(panelSizeChanged, () => {
         ctrl.calculatePanelHeight(panelContainer[0].offsetHeight);
         $timeout(() => {
           resizeScrollableContent();
           ctrl.render();
         });
       });
 
-      ctrl.events.on('view-mode-changed', () => {
+      ctrl.events.on(viewModeChanged, () => {
```
Comment (torkelo) [inline public/app/features/panel/panel_directive.ts]: this is also a PanelEvents event

File: public/app/features/panel/panel_directive.ts [hunk 21]
```diff
@@ -123,7 +125,7 @@ module.directive('grafanaPanel', ($rootScope, $document, $timeout) => {
         }, 10);
       });
 
-      ctrl.events.on('render', () => {
+      ctrl.events.on(panelEventRender, () => {
```
Comment (torkelo) [inline public/app/features/panel/panel_directive.ts]: this is also a PanelEvents event

File: public/app/plugins/panel/alertlist/module.ts [hunk 22]
```diff
@@ -42,8 +42,8 @@ class AlertListPanel extends PanelCtrl {
     super($scope, $injector);
     _.defaults(this.panel, this.panelDefaults);
 
-    this.events.on('init-edit-mode', this.onInitEditMode.bind(this));
-    this.events.on('refresh', this.onRefresh.bind(this));
+    this.events.on(editModeInitialized, this.onInitEditMode.bind(this));
+    this.events.on(clientRefreshed, this.onRefresh.bind(this));
```
Comment (torkelo) [inline public/app/plugins/panel/alertlist/module.ts]: These two are PanelEvents,

File: public/app/plugins/panel/graph/module.ts [hunk 23]
```diff
@@ -146,12 +155,12 @@ class GraphCtrl extends MetricsPanelCtrl {
...
-    this.events.on('data-error', this.onDataError.bind(this));
-    this.events.on('data-snapshot-load', this.onDataSnapshotLoad.bind(this));
-    this.events.on('init-edit-mode', this.onInitEditMode.bind(this));
-    this.events.on('init-panel-actions', this.onInitPanelActions.bind(this));
+    this.events.on(panelEventRender, this.onRender.bind(this));
+    this.events.on(dataFramesReceived, this.onDataFramesReceived.bind(this));
+    this.events.on(PanelEvents.dataReceived, this.onDataError.bind(this));
+    this.events.on(PanelEvents.dataSnapshotLoad, this.onDataSnapshotLoad.bind(this));
+    this.events.on(editModeInitialized, this.onInitEditMode.bind(this));
+    this.events.on(initPanelActions, this.onInitPanelActions.bind(this));
```
Comment (torkelo) [inline public/app/plugins/panel/graph/module.ts]: all these are panel events

**Review by torkelo — COMMENTED** [review]
Still some inconsistency in naming / namespacing that does not feel right. 

Think all events that are published / subscribed on the PanelModel.events should be part of the PanelEvents namespace not just some of the data events. 

For the events in @grafana/data , they should be namespaced under AppEvents and events defined in app/types/events.ts should be namespaced with either "InternalEvents" or "CoreEvents" ?

File: public/app/features/panel/panel_ctrl.ts (hunk 19 above)
Comment (kaydelaney) [reply public/app/features/panel/panel_ctrl.ts]: Mentioned above that since this event's payload depends on `TableModel` and `TimeSeries` it can't be moved to @grafana/ui (where PanelEvents is defined) without moving `TableModel` and `TimeSeries` to @grafana/data

Comment (torkelo) [reply public/app/features/panel/panel_ctrl.ts]: the render payload should be up to each panel so no hard requirement on TableModel | TimeSeries.

**Review by torkelo — APPROVED** [review]
Looks good now! Great job!

File: packages/grafana-ui/src/types/events.ts [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+export interface MenuElement {
+  text: string;
+  click: string;
+  role?: string;
+  shortcut?: string;
+}
+
+/** Events */
+
+export const clientRefreshed = eventFactory('refresh');
```
Comment (torkelo) [inline packages/grafana-ui/src/types/events.ts]: since we have namespacing of these events now. this could just be refresh to have same name as event string

File: packages/grafana-ui/src/types/events.ts [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+export const dataError = eventFactory<DataQueryError>('data-error');
+export const dataReceived = eventFactory<DataQueryResponseData[]>('data-received');
+export const dataSnapshotLoad = eventFactory<DataQueryResponseData[]>('data-snapshot-load');
+export const editModeInitialized = eventFactory('init-edit-mode');
+export const initPanelActions = eventFactory<MenuElement[]>('init-panel-actions');
+export const panelChangeView = eventFactory<PanelChangeViewPayload>('panel-change-view');
+export const panelInitialized = eventFactory('panel-initialized');
+export const panelSizeChanged = eventFactory('panel-size-changed');
+export const panelTeardown = eventFactory('panel-teardown');
+export const removePanel = eventFactory<number>('remove-panel');
```
Comment (torkelo) [inline packages/grafana-ui/src/types/events.ts]: this is not a panel event (not ever emitted on the PanelModel.events), could be moved to the core events
